op-supervisor/supervisor/backend/db/fromda/db.go

```diff
@@ -0,0 +1,235 @@
+package fromda
+
+import (
+	"cmp"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/backend/db/entrydb"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
+type Metrics interface {
+	RecordDBDerivedEntryCount(count int64)
+}
+
+type EntryStore interface {
+	Size() int64
+	LastEntryIdx() entrydb.EntryIdx
+	Read(idx entrydb.EntryIdx) (Entry, error)
+	Append(entries ...Entry) error
+	Truncate(idx entrydb.EntryIdx) error
+	Close() error
+}
+
+// DB implements an append only database for log data and cross-chain dependencies.
+// Each entry is fixed size, and denotes an increment in L1 (derived-from) and/or L2 (derived) block.
+// Data is an append-only log, that can be binary searched for any necessary derivation-link data.
+type DB struct {
+	log    log.Logger
+	m      Metrics
+	store  EntryStore
+	rwLock sync.RWMutex
+}
+
+func NewFromFile(logger log.Logger, m Metrics, path string) (*DB, error) {
+	store, err := entrydb.NewEntryDB[EntryType, Entry, EntryBinary](logger, path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open DB: %w", err)
+	}
+	return NewFromEntryStore(logger, m, store)
+}
+
+func NewFromEntryStore(logger log.Logger, m Metrics, store EntryStore) (*DB, error) {
+	db := &DB{
+		log:   logger,
+		m:     m,
+		store: store,
+	}
+	db.m.RecordDBDerivedEntryCount(db.store.Size())
+	return db, nil
+}
+
+// Rewind to the last entry that was derived from a L1 block with the given block number.
+func (db *DB) Rewind(derivedFrom uint64) error {
+	index, _, err := db.lastDerivedAt(derivedFrom)
+	if err != nil {
+		return fmt.Errorf("failed to find point to rewind to: %w", err)
+	}
+	err = db.store.Truncate(index)
+	if err != nil {
+		return err
+	}
+	db.m.RecordDBDerivedEntryCount(int64(index) + 1)
+	return nil
+}
+
+// Latest returns the last known values:
+// derivedFrom: the L1 block that the L2 block is safe for (not necessarily the first, multiple L2 blocks may be derived from the same L1 block).
+// derived: the L2 block that was derived (not necessarily the first, the L1 block may have been empty and repeated the last safe L2 block).
+func (db *DB) Latest() (derivedFrom types.BlockSeal, derived types.BlockSeal, err error) {
+	db.rwLock.Lock()
+	defer db.rwLock.Unlock()
+	return db.latest()
+}
+
+// latest is like Latest, but without lock, for internal use.
+func (db *DB) latest() (derivedFrom types.BlockSeal, derived types.BlockSeal, err error) {
+	lastIndex := db.store.LastEntryIdx()
+	if lastIndex < 0 {
+		return types.BlockSeal{}, types.BlockSeal{}, entrydb.ErrFuture
+	}
+	last, err := db.readAt(lastIndex)
+	if err != nil {
+		return types.BlockSeal{}, types.BlockSeal{}, fmt.Errorf("failed to read last derivation data: %w", err)
+	}
+	return last.derivedFrom, last.derived, nil
+}
```

Comment:
Suggest making a compound type `L1Derivation` or something to house the two fields. That way it's not on the caller to maintain the relationship between the two `BlockSeal`s.

```go
type Derivation struct {
Derived types.BlockSeal
From types.BlockSeal
}
```

Reply:
I abandoned this since callers may only care about half the output.